salt/modules/virt.py

```diff
@@ -2801,14 +2921,26 @@ def purge(vm_, dirs=False, removables=None, **kwargs):
     for disk in disks:
         if not removables and disks[disk]['type'] in ['cdrom', 'floppy']:
             continue
-        os.remove(disks[disk]['file'])
-        directories.add(os.path.dirname(disks[disk]['file']))
+        elif disks[disk].get('zfs', False):
+            # Sleep a little while to avoid 'dataset is busy'
+            time.sleep(3)
+            fs_name = disks[disk]['file'][len('/dev/zvol/'):]
+            log.info('Destroying VM ZFS volume {0}'.format(fs_name))
+            __salt__['zfs.destroy'](
+                    name=fs_name,
+                    force=True)
+        else:
+            os.remove(disks[disk]['file'])
+            directories.add(os.path.dirname(disks[disk]['file']))
     if dirs:
         for dir_ in directories:
             shutil.rmtree(dir_)
     if getattr(libvirt, 'VIR_DOMAIN_UNDEFINE_NVRAM', False):
         # This one is only in 1.2.8+
-        dom.undefineFlags(libvirt.VIR_DOMAIN_UNDEFINE_NVRAM)
+        try:
+            dom.undefineFlags(libvirt.VIR_DOMAIN_UNDEFINE_NVRAM)
+        except libvirt.libvirtError as ex:
+            dom.undefine()
```

Comment:
That hunk should go in a separate commit since that has nothing to do with bhyve

Reply:
It does. On FreeBSD calling undefineFlags raises `libvirt: Domain Config error : this function is not supported by the connection driver: virDomainUndefineFlags`

Reply:
ok, but I suspect this is not the only driver in that case. I still think it would be better to have it in a separate commit with what you just wrote as commit description to explain the reason of the change